### PR TITLE
[RHCLOUD-35383] Add v2_role FK, indexes, and unique constraint to BindingMapping

### PR DESCRIPTION
**Pull request by dsova06**
## Link(s) to Jira
- [RHCLOUD-35383](https://redhat.atlassian.net/browse/RHCLOUD-35383)

## Description of Intent of Change(s)

**WHAT:** Add a ForeignKey from BindingMapping to RoleV2, compound indexes, and a unique constraint to prevent duplicate binding mappings.

**WHY:** BindingMapping currently stores the V2 role ID buried inside a JSON field (`mappings["role"]["id"]`) and has zero indexes or constraints. This causes inefficient lookups, no database-level prevention of duplicate bindings, and required a complex locking workaround in the dual-write handler to prevent concurrent duplicate creation.

**HOW:**
- Add nullable `v2_role` ForeignKey to BindingMapping pointing to RoleV2 (SET_NULL on delete)
- Add compound index on `(role, resource_type_namespace, resource_type_name, resource_id)` for V1 role queries used in the locking code and workspace service
- Add unique constraint on `(v2_role, resource_type_namespace, resource_type_name, resource_id)` to prevent duplicate bindings per V2 role and resource (merged into schema migration since PostgreSQL excludes NULLs from unique constraints)
- Data migration backfills `v2_role` from `mappings["role"]["id"]` using `values_list` for UUID collection, `itertools.batched` for chunking, `transaction.atomic()` + `select_for_update()` per batch for safe locking, and fallback individual lookups for race condition handling
- Make `v2_role` a required parameter on `for_role_binding()` factory method (callers must explicitly pass it, even if `None`)
- Update `sharedSystemRolesReplicatedRoleBindings` caller to pass `v2_role`
- Update dual-write handler to look up `SeededRoleV2` by UUID for system roles

Two migrations: schema (add field + index + unique constraint) -> data (backfill with batched locking).

SET_NULL is used instead of CASCADE to preserve BindingMappings when a RoleV2 is deleted. The V1 `role` FK already handles cleanup via CASCADE when a V1 role is deleted. SET_NULL allows lazy V2 role re-creation from existing mappings (used by the dual-write handler's custom role migration path). Long-term target is PROTECT (tracked as follow-up).

**Out of scope (follow-up):**
- Switching `on_delete` from SET_NULL to PROTECT (after lazy migration path is removed)
- Making `v2_role` field non-nullable (after backfill completes across all tenants)
- Removing the resource-level locking workaround in `relation_api_dual_write_subject_handler.py`
- Replacing `mappings["role"]["id"]` JSON lookups with `v2_role` FK queries

**Note:** Replaces PR #2906 (closed due to GitHub signed-commits ruleset conflict during rebase).

## Local Testing
- `pipenv run tox -e lint` -- passes
- `pipenv run tox -e py312-fast` -- all 3795 tests pass
- Migrations applied and verified against local PostgreSQL

## Checklist
- [ ] if API spec changes are required, is the spec updated? N/A
- [ ] are there any pre/post merge actions required? No
- [x] are these changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for? N/A
- [x] does this require migration changes?
  - [x] if yes, are they backwards compatible? Yes -- new nullable FK, additive index and constraint
- [ ] is there known, direct impact to dependent teams/components? No

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

[RHCLOUD-35383]: https://redhat.atlassian.net/browse/RHCLOUD-35383?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Add a nullable RoleV2 foreign key and supporting constraints to BindingMapping, backfill it from existing JSON data, and update callers and tests accordingly.

New Features:
- Introduce a nullable v2_role foreign key on BindingMapping linking to RoleV2.
- Enforce a unique BindingMapping per v2_role and resource via a database unique constraint.

Enhancements:
- Add a compound index on BindingMapping for role and resource fields to improve lookup performance.
- Extend BindingMapping.for_role_binding to accept an explicit v2_role parameter and propagate it to created mappings.
- Update dual-write and migration tooling to populate and maintain the new v2_role relationship.

Tests:
- Add tests covering the v2_role unique constraint behavior, NULL handling, on-delete SET_NULL semantics, and the updated for_role_binding factory.
- Adjust existing tests and helpers to pass the new v2_role argument and validate its presence in API responses.

### CODE REVIEW

**Comment by dsova06** [comment]
This replaces PR #2906, which was closed because GitHub's new signed-commits ruleset blocked force-pushes after rebasing onto master -- the rebase surfaced unsigned ancestor commits from the old master history, and the ruleset rejected the entire push. Deleting and recreating the branch was the only workaround, which closed the original PR.

All review feedback from #2906 has been incorporated (batched locking, `itertools.batched`, `values_list`, race condition handling, migration consolidation, required `v2_role` parameter). The code is identical to the last force-push on #2906, plus a rebase onto current master with renumbered migrations (0089/0090).

@randomnetcat -- requesting your re-review. All your previous comments were addressed and the threads resolved on #2906.

File: rbac/management/migrations/0090_backfill_bindingmapping_v2_role.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+def backfill_v2_role(apps, schema_editor):
+    """Backfill v2_role FK from mappings JSON field."""
+    BindingMapping = apps.get_model("management", "BindingMapping")
+    RoleV2 = apps.get_model("management", "RoleV2")
+
+    needed_uuids = set(
+        BindingMapping.objects.filter(v2_role__isnull=True)
+        .exclude(mappings__role__id=None)
+        .values_list("mappings__role__id", flat=True)
```
Comment (randomnetcat) [inline rbac/management/migrations/0090_backfill_bindingmapping_v2_role.py:14]: `.distinct()`?

File: rbac/management/migrations/0090_backfill_bindingmapping_v2_role.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    )
+
+    v2_role_lookup = {str(r.uuid): r.pk for r in RoleV2.objects.filter(uuid__in=needed_uuids).only("id", "uuid")}
+
+    unprocessed_pks = list(BindingMapping.objects.filter(v2_role__isnull=True).values_list("pk", flat=True))
```
Comment (randomnetcat) [inline rbac/management/migrations/0090_backfill_bindingmapping_v2_role.py]: Might as well use `.iterator()` if we're already batching them locally.

**Review by randomnetcat — COMMENTED** [review]
We should also:
* Check that `role_v2` is actually set on `BindingMapping`s created during dual-write.
* Update `assert_v1_v2_locally_consistent` to check that `v2_role` is consistent (both with the corresponding `RoleBinding` and with `mappings["role"]["id"]`.

File: rbac/management/migrations/0090_backfill_bindingmapping_v2_role.py (hunk 1 above)
Comment (dsova06) [reply rbac/management/migrations/0090_backfill_bindingmapping_v2_role.py:14]: Fixed: added `.distinct()` to deduplicate UUIDs at the database level.

File: rbac/management/migrations/0090_backfill_bindingmapping_v2_role.py (hunk 2 above)
Comment (dsova06) [reply rbac/management/migrations/0090_backfill_bindingmapping_v2_role.py]: Fixed: switched to `.iterator()` instead of `list()` so PKs stream through `itertools.batched` without loading all into memory at once.

**Review by randomnetcat — COMMENTED** [review]
What's here looks fine, but see [my previous comments](https://github.com/project-kessel/insights-rbac/pull/3004#pullrequestreview-4477872820) for some things we should add.

**Comment by dsova06** [comment]
Addressed both items:

1. **`assert_v1_v2_locally_consistent` updated** -- `_assert_binding_mapping_consistent` now checks that `v2_role` is set, matches `binding.role` (the RoleBinding's V2 role), and that `v2_role.uuid` matches `mappings["role"]["id"]`. Also added `select_related("v2_role")` to avoid N+1 queries.

2. **Dual-write test coverage** -- `expect_binding_present` in `test_dual_write.py` now asserts `v2_role` is set and matches the expected role UUID. Since `assert_v1_v2_locally_consistent` is called in `tearDown` of role view tests and throughout the dual-write and migration tests, the new assertions verify `v2_role` is populated on every BindingMapping created through any dual-write path.

**Review by randomnetcat — APPROVED** [review]

**Comment by vbelchio** [comment]
QE: NA
